### DISCUSSION TITLE
Integrating with C# Project

### DISCUSSION THREAD

**Discussion by RobDM** [dialogic-godot/dialogic#2157] [Q&A] [answered]
Hi there,

I'm completely new to Dialogic, and I'm having a rough time trying to integrate it into an existing C# project that I've been working on for the past few months.

I downloaded the latest version a few days ago, and followed the instructions to install it in the /addons folder.  The plugin UI shows up in the Godot editor and appears to be set up correctly, as best I can tell.  I also got the C# wrapper from [this thread](https://github.com/dialogic-godot/dialogic/discussions/1776) - thanks QueenOfSquiggles!

1. Every few times I start the project from the Godot IDE, there is a break on DialogicGameHandler.gd line 29 (`var current_timeline: DialogicTimeline = null`) and if I resume execution, the dialog layout doesn't show up at all when I try to trigger it.  It seems to be failing to load or initialize.
2. If I run the project in debug mode from VSCode, the Dialogic autoload doesn't happen at all, 100% of the time.
3. When I can get the dialog layout to show up, there is a null reference break each time a text box is shown on animations.gd line 78 (`animation_system.call(&'start_animating')`) because animation_system is null.  If I resume execution, it seems to continue working fine.
4. I created a test timeline with two characters on opposite sides of the screen (positions 0 and 4).  The one on the left has a "slide in left" entry animation, and the one on the right has a "slide in right" animation.  But when the timeline runs, the character on the left appears instantly, and the character on the right slides across the entire screen from the left and off the other side, before sliding back in from the right.  It's like both character's animations are being played in sequence on the second character.

Does anyone know how I can solve any of the above issues?

Thank you!

**Comment by CakeVR** [comment]
Hello, there is no official Dialogic C# Wrapper as of right now, you will have to make your own little wrapper or call GDScript: https://docs.godotengine.org/en/stable/tutorials/scripting/cross_language_scripting.html

The Dialogic C# Wrapper you are referencing looks quite outdated and likely causes you trouble.

**Reply by RobDM** [reply]
Thanks for the quick reply CakeVR!  Actually I was experiencing most of the issues above before I found the unofficial (and outdated) wrapper, and they persisted even after I introduced it.  So the wrapper may be outdated, but the problems I'm having seem to be with the addon itself.  Like the first issue will happen even if I don't invoke any dialog code, just having the plugin present and autoloaded causes it.

I will try from a fresh project and see if any of the issues are still reproducible.

**Reply by RobDM** [reply]
From adding Dialogic to a fresh project I found out:

- Problem 1 and 3 are no longer reproducible.  They only seem to happen in my existing project.  I have absolutely no clue why though.
- Problem 2 is 100% reproducible.  The autoload is missed somehow when running the project in debug mode from VSCode.  I have no idea if this is the case in general with autoloads, or if it's specific to Dialogic.
- Problem 4 was not reproducible in the fresh project.  Even without checking "Await end" on either of the character Join events, both animated in correctly.  Thought the "Slide from left" and "Slide from right" transitions seem to have a slight overshot that is not so slight on lower resolution projects (480x240).

I guess what's left is for me to try setting Dialogic up again from scratch in my project, in case some of the code or assets were corrupted somehow.

**Reply by CakeVR** [reply]
I would try the Autoload issue with a different autoload too, it works fine for me with VSCode, it may be a C# issue.

**Reply by RobDM** [reply]
Alright, after some more messing around here's where I am.

- I completely removed Dialogic from my project and tried to integrate it again from scratch.  Same result as above.
- I disabled the autoload and manually added a node to my game hierarchy, and attached the DialogicGameHandler.gd script to it.  This solved Problem 2 from above, Dialogic will load even when the project is started in debug mode from VSCode.  So this is progress.
- Now when I trigger a timeline to start, Dialogic blows up in history_layer.gd with **Identifier "Dialogic" not declared in the current scope** for line 56: `Dialogic.History.open_requested.connect(_on_show_history_pressed)`.  The dialog doesn't show up at all when I resume execution, so now it's completely broken.

One step forward, 2 steps back. :/ These all seem to be loading issues, like parts of the Dialogic scripts are missing.  I was able to get it working in a C# project with the outdated wrapper in my message above, so it has to be something specific to my project.  But I'm not doing anything out of the ordinary for the game start-up, so it's baffling to me.

What I may try doing next is start a fresh project with Dialogic working and try to integrate parts of my game piece by piece until I find where it breaks.  It's a little extreme but the only thing I can think of to move forward.

**Reply by CakeVR** [reply]
This seems to be an issue where Autoloads don't work in C# projects properly.
Dialogic must be autoloaded, otherwise nothing in Dialogic will work, see the many errors you encountered.

You should try what I suggested before. Make your own Autoload and try using that, see if it breaks in the same way.

**Reply by RobDM** [reply, marked as answer]
Yep you're right, I tried adding my own autoload and it doesn't load when started in debug mode from VSCode, but it does load when run from within the Godot editor.  :/

Yesterday I was going through the process of starting with a minimal project with Dialogic working, and integrating parts of my project bit by bit until it broke.  I was able to find the tipping point, though it's not nearly as predictable as I'd hoped.  I got to the point where I brought all of my C# scripts into the project and Dialogic was still working.  Then after a few seemingly harmless changes to one of my scripts, Dialogic started to fail to load again.  And even when I rolled back those changes, it continued to not load.  It took restarting the Godot editor to get it back to a good state.

So I guess the answer is that the autoload problems start happening randomly any time the C# code is rebuilt, and the more files you have, the more likely this is to happen.  This somehow puts Dialogic into a bad state for the rest of the session, and the editor has to be restarted (and I would bet this would be the case with just about any Godot addon in a C# project).  Although Godot 4 supports C#, they've clearly not worked out the kinks of mixing scripts together yet, beyond the simple scenarios.

Unfortunately for me, this means I won't be able to use Dialogic in my main project (for now), which is a huge bummer because I really love the UI.  However, I think it can still be useful for helping me to prototype my dialog before implementing it in my main project....I just have to keep it in a separate project.

Anyway, thanks for helping me troubleshoot!
